clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -0,0 +1,182 @@
+
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setResourceName(requestParts[4])
+        .setPartition(Integer.parseInt(requestParts[5]))
+        .setKeyString(requestParts[6])
+        .setIsBatchRequest(false)
+        .build();
+
+    VeniceReadServiceGrpc.VeniceReadServiceStub clientStub = getStub(channel);
+    GrpcTransportClientCallback callback = new GrpcTransportClientCallback(clientStub, request);
+
+    return callback.get();
+  }
+
+  @Override
+  public CompletableFuture<TransportClientResponse> post(
+      String requestPath,
+      Map<String, String> headers,
+      byte[] requestBody) {
+    String[] requestParts = requestPath.split("/");
+    if (!requestParts[2].equals(STORAGE_ACTION)) {
+      LOGGER.debug(
+          "performing unsupported gRPC transport client action ({}), passing request to R2 client",
+          requestParts[2]);
+      return r2TransportClient.post(requestPath, headers, requestBody);
+    }
+
+    ManagedChannel channel = getChannel(requestParts[2]);
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setResourceName(requestParts[4])
+        .setKeyBytes(ByteString.copyFrom(requestBody))
+        .setIsBatchRequest(true)
+        .build();
+
+    VeniceReadServiceGrpc.VeniceReadServiceStub clientStub = getStub(channel);
+    GrpcTransportClientCallback callback = new GrpcTransportClientCallback(clientStub, request);
+    return callback.post();
+  }
+
+  @Override
+  public void close() {
+    for (Map.Entry<String, ManagedChannel> entry: serverGrpcChannels.entrySet()) {
+      entry.getValue().shutdown();
+    }
+  }
+
+  private static class GrpcTransportClientCallback {
+    // start exception handling
+    private final CompletableFuture<TransportClientResponse> valueFuture;
+    private final VeniceReadServiceGrpc.VeniceReadServiceStub clientStub;
+    private final VeniceClientRequest request;
+
+    public GrpcTransportClientCallback(
+        VeniceReadServiceGrpc.VeniceReadServiceStub clientStub,
+        VeniceClientRequest request) {
+      this.clientStub = clientStub;
+      this.request = request;
+      this.valueFuture = new CompletableFuture<>();
+    }
+
+    public CompletableFuture<TransportClientResponse> get() {
+      if (request.getIsBatchRequest()) {
+        throw new UnsupportedOperationException("Not a single get request, use batchGet() instead");
+      }
+      clientStub.get(request, new StreamObserver<VeniceServerResponse>() {
+        @Override
+        public void onNext(VeniceServerResponse value) {
+          valueFuture.complete(
+              new TransportClientResponse(
+                  value.getSchemaId(),
+                  CompressionStrategy.valueOf(value.getCompressionStrategy()),
+                  value.getData().toByteArray()));
+        }
+
+        @Override
+        public void onError(Throwable t) {
+          LOGGER.error("Error in gRPC request", t);
```

Comment:
Normally, we don't want to log the exception here since normally application will catch the exception anyway.